pkg/kv/kvserver/loqrecovery/collect.go

```diff
@@ -45,14 +51,22 @@ func CollectReplicaInfo(
 			if err != nil {
 				return err
 			}
+			// Check raft log for un-applied range descriptor changes. We start from
+			// the position after applied (inclusive) and stop at position after
+			// committed (exclusive) - hence +1 on index values.
+			hasUnappliedDesc, err := checkUncommittedDescriptorsInRaftLog(desc.RangeID, desc.StartKey,
+				rstate.RaftAppliedIndex+1, hstate.Commit+1, reader)
```

Comment:
I think you'd want to check up to the end of the log here (i.e. LastIndex+1 as opposed to Commit+1). Otherwise, we have to also truncate the log because any conf change in the "uncommitted" tail would become committed on the survivor and cause the very same problems we're trying to avoid. Also, note that even though not "known committed", such an entry may in fact have committed, meaning that the meta2 record might be the "new" descriptor - however I think this is ok since the next replication change will patch up the descriptor. For splits and merges it's a whole different story though, it will require a game of non-local analysis and the ability to apply splits "offline" to get out of the situation.
For now I strongly prefer we just catch all mutations to the descriptor and bail.

Reply:
So all the detection was moved into checking commit triggers and bailing in presence of anything but removing other replicas.